File: 0238-product-of-array-except-self/0238-product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        /*
        Create an array ans and initializeall of it's elements to 1 and a variable temp = 1.
        Traverse the array from start to end.
        For every index i update ans[i] as ans[i] = temp and temp = temp * nums[i], i.e store the product upto i-1 index from the start of array.
        Now again, temp = 1 and traverse the array in the reverse direction.
        For every index i update ans[i] as ans[i] = ans[i] * temp and *temp = temp nums[i], i.e multiply with the product upto i+1 index from the end of array.

        Time Complexity : O(n)
        Auxiliary Space : O(1)
        */
        int n=nums.size(), temp=1;
        vector<int> ans(n, 1);
        
        for(int i=0;i<n;i++){
            ans[i] *= temp;
            temp *= nums[i];
        }
        temp = 1;
        for(int i=n-1;i>=0;i--){
            ans[i] *= temp;
            temp *= nums[i];
        }
        return ans;
    }
};

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.cs
public class Solution {
    public int[] ProductExceptSelf(int[] nums) {
        /*
        Use two arrays to calculate the left of the products and right of the products of the arry
        First, the left array iterates through the current index's left and calculate the product
        At the second loop iterate through right array and calculate the product of the current product
        Multiply the left and right array elements

        Time Complexity : O(n)
        Auxiliary Space : O(n)
        */

         List<int> left = new List<int>();
        List<int> right = new List<int>();

        List<int> result = new List<int>();


        int leftMultp = 1;
        int rightMultp = 1;

        for(int i = 0; i < nums.Length; i++)
        {
            if(i == 0)
                left.Add(1);
            
            leftMultp *= nums[i];

            left.Add(leftMultp);
        }

        for(int i = nums.Length - 1; i >= 0; i--)
        {
            if(i == nums.Length - 1)
                right.Add(1);
            
            rightMultp *= nums[i];
            right.Add(rightMultp);

        }

        for(int i = 0; i < nums.Length; i++)
            result.Add(left[i] * right[nums.Length - 1 - i]);

        return result.ToArray();
    }
}